mitodl/open-discussions
# Allow programs, videos, and user lists to be added as UserList items

#### Pre-Flight checklist

#### What are the relevant tickets?
Fixes #2344

#### What's this PR do?
- Makes `GenericForeignKeySerializer` handle programs, videos, and lists
- Replaces `FavoriteItemContentSerializer` with the above
- Changes verbose name of `UserList` to `userlist` (matching name)
- Adds a doc summarizing much of the code that needs to be changed when new learning resource types are added.

#### How should this be manually tested?
- Favorite and unfavorite some courses, programs, videos, etc. 
- Go to `http://od.odl.local:8063/api/v0/userlists/<user_list_id>` where user_list_id is the id of an existing `UserList` (create one if necessary).
- PATCH the following, it should succeed:
```python
{ "items": 
  [
    {
         "content_type": "video",
        "object_id": <id of existing video>
    },
    {
         "content_type": "program",
        "object_id": <id of existing program>
    },
    {
         "content_type": "userlist",
        "object_id": <id of one of your other lists>
    }
  ]
}
```
- Recreate your index, check that search still works for user lists and learning paths
